src/api.cpp

```diff
@@ -0,0 +1,301 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2021 YADRO
+
+#include <errno.h>
+#include <stdio.h>
+
+#include <libobmcsession/obmcsession.hpp>
+#include <manager.hpp>
+#include <session.hpp>
+
+using namespace obmc::session;
+static SessionManagerPtr managerPtr;
+static std::unique_ptr<sdbusplus::bus::bus> dbus;
+
+int obmcsesManagerInit(sd_bus* bus, const char* slug, uint8_t type);
+int obmcsesManagerInitAsio(sdbusplus::bus::bus& bus, const std::string& slug,
+                           uint8_t type);
+int obmcsesCreate(const char* userName, uint32_t remoteAddress,
+                  obmcsesSessionId* sessionId);
+
+int obmcsesCreateTransaction(obmcsesSessionId* sessionId);
+
+int obmcsesCreateTransactionWithCleanup(obmcsesCleanupFn cleanupFn,
+                                        obmcsesSessionId* sessionId);
+
+int obmcsesCreateTransactionWithFCleanup(SessionCleanupFn&& cleanupFn,
+                                         obmcsesSessionId* sessionId);
+
+int obmcsesCreateWithCleanup(const char* userName, uint32_t remoteAddress,
+                             obmcsesCleanupFn cleanupFn,
+                             obmcsesSessionId* sessionId);
+int obmcsesCreateWithFCleanup(const std::string& userName,
+                              uint32_t remoteAddress,
+                              SessionCleanupFn&& cleanupFn,
+                              obmcsesSessionId* sessionId);
+int obmcsesCommitSessionBuild(const char* username, uint32_t remoteIPAddr);
+int obmcsesCommitSessionBuildRemote(sd_bus* bus, const char* slug,
+                                    const char* username,
+                                    uint32_t remoteIPAddr);
+int obmcsesCommitSessionRemoteAsio(sdbusplus::bus::bus& bus,
+                                   const std::string& slug,
+                                   const std::string& username,
+                                   uint32_t remoteIPAddr);
+bool obmcsesRemove(obmcsesSessionId sessionId);
+unsigned int obmcsesRemoveAllByUser(const char* userName);
+unsigned int obmcsesRemoveAllByAddress(uint32_t remoteAddress);
+unsigned int obmcsesRemoveAllBytType(uint8_t type);
+unsigned int obmcsesRemoveAll();
+bool obmcsesIsSessionBuildPending();
+void obmcsesResetPendginSessilBuild();
+
+int obmcsesManagerInit(sd_bus* bus, const char* slug, uint8_t type)
+{
+    if (dbus || managerPtr)
+    {
+        return EEXIST;
+    }
+    if (!bus)
+    {
+        return EINVAL;
+    }
+    try
+    {
+        dbus = std::make_unique<sdbusplus::bus::bus>(bus);
```

Comment:
As the separate C++ API is excess, I suggest to pass the raw `sd_bus` pointer to the SessionManager constructor.
In addition, `sdbusplus::bus` is kind of smart pointer and it doesn't require the extra wrapping into `std::uniq_ptr`.

Reply:
I've still beliave the Cpp impelentation of serveral API-functions is required to cleanly provide sdbusplus handles.

Reply:
Ok, the C++ API is also required. But keeping the bus object wrapped into `std::uniq_ptr` seems to me like overhead.
This still could be a member of `SessionManager` that might have two constructors.
The first gets the raw pointer and create `sdbusplus::bus` object from it,
the second just copies the bus object from C++ app.

Reply:
Yes, it makes sense. I'll add another constructor with sd_bus. Thanks for the proposal.